### PR TITLE
Bahamas Client Challenge

### PR DESCRIPTION
**Pull request by GBTelles1**

### CODE REVIEW

File: 1-development/bahamasclient/.ruby-version [hunk 1]
```diff
@@ -0,0 +1 @@
+3.0.3
```
Comment (goncalopc96) [inline 1-development/bahamasclient/.ruby-version:1]: Why didn't you use the newest ruby version?

File: 1-development/bahamasclient/test/controllers/bahamas_clients_controller_test.rb [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+class BahamasClientsControllerTest < ActionDispatch::IntegrationTest
+  test "should get store" do
+    get bahamas_clients_store_url
+    assert_response :success
+  end
+
+  test "should get retrieve" do
+    get bahamas_clients_retrieve_url
+    assert_response :success
+  end
```
Comment (goncalopc96) [inline 1-development/bahamasclient/test/controllers/bahamas_clients_controller_test.rb:12]: What happens if something goes wrong in the actions of the controller? Will it raise specific errors for some possible scenarios? for example while saving or searching in the database, making a request, and like you have in the comments, if it has a ParameterMissing error, will it be correctly raised and send the correct error code and message?

File: 1-development/bahamasclient/app/models/client.rb [hunk 3]
```diff
@@ -0,0 +1,6 @@
+class Client < ApplicationRecord
+  has_and_belongs_to_many :invoices
```
Comment (jeandias) [inline 1-development/bahamasclient/app/models/client.rb:2]: In our current model, a client has many invoices and an invoice belongs to a single client.
What would be the approach based on the current model?

File: 1-development/bahamasclient/test/controllers/bahamas_clients_controller_test.rb [hunk 4]
```diff
@@ -0,0 +1,13 @@
+require "test_helper"
+
+class BahamasClientsControllerTest < ActionDispatch::IntegrationTest
+  test "should get store" do
```
Comment (jeandias) [inline 1-development/bahamasclient/test/controllers/bahamas_clients_controller_test.rb:4]: README talks about codes 200, 201 and 422, so shouldn't the tests reflect that?

File: 1-development/bahamasclient/README.md [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+
+```json
+{
+  "id": "1",
+  "fiscal_id": "999999999",
+  "name": "Bob",
+  "email": "bob@bob.com",
+  "created_at": "2024-08-11T12:34:56.789Z",
+  "updated_at": "2024-08-11T12:34:56.789Z",
+  "invoices": []
```
Comment (jeandias) [inline 1-development/bahamasclient/README.md:69]: Shouldn't the associated invoice be here?

File: 1-development/bahamasclient/app/controllers/bahamas_clients_controller.rb [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+    # Check if a client with the same fiscal_id already exists
+    client = Client.find_by(fiscal_id: fiscal_id)
+
+    if client
+      # If client exists, associate with the invoice
+      invoice.clients << client unless invoice.clients.include?(client)
+      render json: { message: "Client #{client.name} has been associated with the invoice #{invoice.id}." }, status: :ok
+    else
+      # Create a new client and associate with the invoice
+      client = invoice.clients.create(
```
Comment (jeandias) [inline 1-development/bahamasclient/app/controllers/bahamas_clients_controller.rb:21]: Wouldn't it be more practical and safer to implement the client_params method?

File: 1-development/bahamasclient/app/controllers/bahamas_clients_controller.rb [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+        render json: { client: client, bahamas_service_response: bahamas_service_response }, status: :created
+      else
+        render json: client.errors, status: :unprocessable_entity
+      end
+    end
+  end
+
+  # GET /retrieve-bahamas-client/:invoice_id
+  def retrieve
+    invoice = Invoice.find_by(id: params[:invoice_id])
```
Comment (jeandias) [inline 1-development/bahamasclient/app/controllers/bahamas_clients_controller.rb:40]: Why not implement the set_invoice method?

**Review by goncalopc96 — COMMENTED** [review]
Hi Gabriel! Hope you enjoyed this challenge! I am one of the developers of InvoiceXpress and I will be making a few commentaries, feel free to respond sincerely :)

**Pull request closed without merging**